matrix-room.c

```diff
@@ -1499,3 +1502,154 @@ void matrix_room_send_message(PurpleConversation *conv, const gchar *message)
+	gint index, length;
+	gint num_joined_members;
+
+    chunk_obj = matrix_json_node_get_object(chunk);
+	if (chunk_obj == NULL) {
+		return;
+	}
+	
+	room_id = matrix_json_object_get_string_member(chunk_obj, "room_id");
+	room = purple_roomlist_room_new(PURPLE_ROOMLIST_ROOMTYPE_ROOM, room_id, NULL);
+	
+	purple_roomlist_room_add_field(roomlist, room, room_id);
+	
+	aliases = matrix_json_object_get_array_member(chunk_obj, "aliases");
+	if (aliases != NULL) {
+		aliases_str = g_string_new(NULL);
+		length = json_array_get_length(aliases);
+		for(index = 0; index < length; index++) {
+			const gchar *alias = json_array_get_string_element(aliases, index);
+			if (index > 0) {
+				aliases_str = g_string_append(aliases_str, ", ");
+			}
+			aliases_str = g_string_append(aliases_str, alias);
+		}
+	}
+  if (aliases_str == 0) {
+    aliases_str = g_string_new(NULL);
+    const gchar * alias = matrix_json_object_get_string_member(chunk_obj, "canonical_alias");
+    if (alias)
+      aliases_str = g_string_append(aliases_str, alias);
+  }
+	purple_roomlist_room_add_field(roomlist, room, aliases_str->str);
+	g_string_free(aliases_str, TRUE);
+	
+	num_joined_members = matrix_json_object_get_int_member(chunk_obj, "num_joined_members");
+	purple_roomlist_room_add_field(roomlist, room, GINT_TO_POINTER(num_joined_members));
+	
+	purple_roomlist_room_add_field(roomlist, room, matrix_json_object_get_string_member(chunk_obj, "name"));
+	purple_roomlist_room_add_field(roomlist, room, matrix_json_object_get_string_member(chunk_obj, "topic"));
+	
+  RoomListAddThread * t = g_malloc0(sizeof(RoomListAddThread));
+  t->roomlist = roomlist;
+  t->room = room;
+  g_mutex_lock(&t->mutex);
+  g_idle_add((GSourceFunc)_roomlist_main_add, t);
+  g_cond_wait(&t->cond, &t->mutex);
+  g_mutex_unlock(&t->mutex);
+  g_free(t);
+  usleep(10 * 1000);
+}
+
+typedef struct _RoomListThread {
+  JsonNode *json_root;
+  JsonObject *rooms;
+  JsonArray *chunk_array;
+  PurpleRoomlist *roomlist;
+} RoomListThread;
+
+static void _roomlist_foreach_thread(RoomListThread* p) {
+  json_array_foreach_element(p->chunk_array, _roomlist_parse_chunk, p->roomlist);
+  json_node_free(p->json_root);
+  g_free(p);
+  purple_roomlist_set_in_progress(p->roomlist, FALSE);
+  pthread_exit(0);
+}
+
+static void _roomlist_got_list(MatrixConnectionData *conn,
+        gpointer user_data, JsonNode *json_root, const char *body, size_t body_len, const char *content_type)
+{
+	PurpleRoomlist *roomlist = user_data;
+	JsonObject *rooms;
+	JsonArray *chunk_array;
+	
+	if (json_root == NULL)
+		return;
+	rooms = matrix_json_node_get_object(json_root);
+	chunk_array = matrix_json_object_get_array_member(rooms, "chunk");
+	
+  RoomListThread * t = g_malloc0(sizeof(RoomListThread));
+  t->json_root = json_node_copy(json_root);
+  t->rooms = rooms;
+  t->chunk_array = chunk_array;
+  t->roomlist = roomlist;
+  pthread_create((pthread_t*)&t->roomlist->proto_data, 0, (void * (*)(void *))_roomlist_foreach_thread, t);
```

Comment:
libpurple is not thread safe, so you shouldn't call any purple_ functions from another thread

Reply:
What method are taking about? purple_roomlist_room_add_field - safe to call from another thread.